File: src/main/scala/jk_5/nailed/worldedit/WorldEditPlayer.scala
package jk_5.nailed.worldedit

import com.google.common.base.Preconditions.checkNotNull
import com.sk89q.worldedit.entity.BaseEntity
import com.sk89q.worldedit.extension.platform.AbstractPlayerActor
import com.sk89q.worldedit.extent.inventory.BlockBag
import com.sk89q.worldedit.internal.LocalWorldAdapter
import com.sk89q.worldedit.internal.cui.CUIEvent
import com.sk89q.worldedit.util.Location
import com.sk89q.worldedit.world.World
import com.sk89q.worldedit.{LocalSession, Vector, WorldEdit, WorldVector}
import jk_5.nailed.api.chat.{ChatColor, ComponentBuilder}
import jk_5.nailed.api.player.Player
import jk_5.nailed.server.NailedServer
import jk_5.nailed.server.player.NailedPlayer
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.item.{Item, ItemStack}

/**
 * No description given
 *
 * @author jk-5
 */
object WorldEditPlayer {
  def wrap(player: EntityPlayerMP): WorldEditPlayer = {
    checkNotNull(player)
    new WorldEditPlayer(NailedServer.getPlayerFromEntity(player))
  }

  def getSession(player: EntityPlayerMP): LocalSession = {
    checkNotNull(player)
    WorldEdit.getInstance.getSessionManager.get(wrap(player))
  }
}

class WorldEditPlayer(p: Player) extends AbstractPlayerActor {
  val np = p.asInstanceOf[NailedPlayer]

  override def getUniqueId = p.getUniqueId
  override def getName = np.getName
  override def getState: BaseEntity = throw new UnsupportedOperationException("Cannot create a state from this object")

  override def getItemInHand: Int = {
    val is = np.getEntity.getCurrentEquippedItem
    if (is == null) 0 else Item.getIdFromItem(is.getItem)
  }

  override def getLocation: Location = {
    val loc = p.getLocation
    val position: Vector = new Vector(loc.getX, loc.getY, loc.getZ)
    new Location(getWorld, position, loc.getYaw, loc.getPitch)
  }

  override def getPosition: WorldVector = {
    val loc = p.getLocation
    new WorldVector(LocalWorldAdapter.adapt(WorldEditWorld.getWorld(np.getEntity.worldObj)), loc.getX, loc.getY, loc.getZ)
  }

  override def getWorld: World = WorldEditWorld.getWorld(np.getEntity.worldObj)
  override def getPitch = np.getEntity.rotationPitch
  override def getYaw = np.getEntity.rotationYaw

  override def giveItem(typ: Int, amt: Int) {
    np.getEntity.inventory.addItemStackToInventory(new ItemStack(Item.getItemById(typ), amt, 0))
  }

  override def dispatchCUIEvent(event: CUIEvent){
    //TODO
    /*String[] params = event.getParameters();
    String send = event.getTypeId();
    if(params.length > 0){
        send = send + "|" + StringUtil.joinString(params, "|");
    }
    Packet250CustomPayload packet = new Packet250CustomPayload(ForgeWorldEdit.CUI_PLUGIN_CHANNEL, send.getBytes(WECUIPacketHandler.UTF_8_CHARSET));
    this.player.playerNetServerHandler.sendPacketToPlayer(packet);*/
  }

  override def printRaw(msg: String){
    msg.split("\n").foreach(p => this.np.sendMessage(new ComponentBuilder(p).create))
  }

  override def printDebug(msg: String){
    msg.split("\n").foreach(p => this.np.sendMessage(new ComponentBuilder(p).color(ChatColor.GRAY).create))
  }

  override def print(msg: String){
    msg.split("\n").foreach(p => this.np.sendMessage(new ComponentBuilder(p).color(ChatColor.LIGHT_PURPLE).create))
  }

  override def printError(msg: String){
    msg.split("\n").foreach(p => this.np.sendMessage(new ComponentBuilder(p).color(ChatColor.RED).create))
  }

  override def setPosition(pos: Vector, pitch: Float, yaw: Float) {
    this.np.netHandler.setPlayerLocation(pos.getX, pos.getY, pos.getZ, pitch, yaw)
  }

  override def getGroups = new Array[String](0)
  override def getInventoryBlockBag: BlockBag = null
  override def hasPermission(perm: String) = p.hasPermission(perm)
  override def getFacet[T](cls: Class[_ <: T]): T = null.asInstanceOf[T]
  override def getSessionKey = new DefaultSessionKey(np)
}

File: src/main/java/jk_5/nailed/plugins/worldedit/WorldEditPlayer.java
package jk_5.nailed.plugins.worldedit;

import com.google.common.base.Charsets;
import com.sk89q.util.StringUtil;
import com.sk89q.worldedit.Vector;
import com.sk89q.worldedit.WorldVector;
import com.sk89q.worldedit.entity.BaseEntity;
import com.sk89q.worldedit.extension.platform.AbstractPlayerActor;
import com.sk89q.worldedit.extent.inventory.BlockBag;
import com.sk89q.worldedit.internal.LocalWorldAdapter;
import com.sk89q.worldedit.internal.cui.CUIEvent;
import com.sk89q.worldedit.session.SessionKey;
import com.sk89q.worldedit.util.Location;
import jk_5.nailed.api.chat.ChatColor;
import jk_5.nailed.api.chat.TextComponent;
import jk_5.nailed.server.NailedPlatform;
import jk_5.nailed.server.player.NailedPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;

import javax.annotation.Nullable;
import java.util.UUID;

public class WorldEditPlayer extends AbstractPlayerActor {

    private final NailedPlayer player;
    private final EntityPlayerMP playerEntity;

    protected WorldEditPlayer(NailedPlayer player) {
        this.player = player;
        this.playerEntity = player.getEntity();
        ThreadSafeCache.getInstance().getOnlineIds().add(getUniqueId());
    }

    @Override
    public UUID getUniqueId() {
        return player.getUniqueId();
    }

    @Override
    public int getItemInHand() {
        ItemStack is = this.playerEntity.getCurrentEquippedItem();
        return is == null ? 0 : Item.getIdFromItem(is.getItem());
    }

    @Override
    public String getName() {
        return player.getName();
    }

    @Override
    public BaseEntity getState() {
        throw new UnsupportedOperationException("Cannot create a state from this object");
    }

    @Override
    public Location getLocation() {
        Vector position = new Vector(this.playerEntity.posX, this.playerEntity.posY, this.playerEntity.posZ);
        return new Location(
                WorldEditWorld.wrap(this.playerEntity.worldObj),
                position,
                this.playerEntity.cameraYaw,
                this.playerEntity.cameraPitch);
    }

    @Override
    public WorldVector getPosition() {
        return new WorldVector(LocalWorldAdapter.adapt(WorldEditWorld.wrap(this.playerEntity.worldObj)), this.playerEntity.posX, this.playerEntity.posY, this.playerEntity.posZ);
    }

    @Override
    public com.sk89q.worldedit.world.World getWorld() {
        return WorldEditWorld.wrap(this.playerEntity.worldObj);
    }

    @Override
    public double getPitch() {
        return this.playerEntity.rotationPitch;
    }

    @Override
    public double getYaw() {
        return this.playerEntity.rotationYaw;
    }

    @Override
    public void giveItem(int type, int amt) {
        this.playerEntity.inventory.addItemStackToInventory(new ItemStack(Item.getItemById(type), amt, 0));
    }

    @Override
    public void dispatchCUIEvent(CUIEvent event) {
        String[] params = event.getParameters();
        String send = event.getTypeId();
        if (params.length > 0) {
            send = send + "|" + StringUtil.joinString(params, "|");
        }
        player.sendPluginMessage(NailedWorldEditPlugin.instance().identifier, "WECUI", send.getBytes(Charsets.UTF_8));
    }

    @Override
    public void printRaw(String msg) {
        for (String part : msg.split("\n")) {
            this.player.sendMessage(new TextComponent(part));
        }
    }

    @Override
    public void printDebug(String msg) {
        for (String part : msg.split("\n")) {
            TextComponent comp = new TextComponent(part);
            comp.setColor(ChatColor.GRAY);
            this.player.sendMessage(comp);
        }
    }

    @Override
    public void print(String msg) {
        for (String part : msg.split("\n")) {
            TextComponent comp = new TextComponent(part);
            comp.setColor(ChatColor.LIGHT_PURPLE);
            this.player.sendMessage(comp);
        }
    }

    @Override
    public void printError(String msg) {
        for (String part : msg.split("\n")) {
            TextComponent comp = new TextComponent(part);
            comp.setColor(ChatColor.RED);
            this.player.sendMessage(comp);
        }
    }

    @Override
    public void setPosition(Vector pos, float pitch, float yaw) {
        this.playerEntity.playerNetServerHandler.setPlayerLocation(pos.getX(), pos.getY(), pos.getZ(), pitch, yaw);
    }

    @Override
    public String[] getGroups() {
        return new String[]{}; // WorldEditMod.inst.getPermissionsResolver().getGroups(this.playerEntity.username);
    }

    @Override
    public BlockBag getInventoryBlockBag() {
        return null;
    }

    @Override
    public boolean hasPermission(String perm) {
        //return ForgeWorldEdit.inst.getPermissionsProvider().hasPermission(playerEntity, perm);
        return true; //TODO: permissions
    }

    @Nullable
    @Override
    public <T> T getFacet(Class<? extends T> cls) {
        return null;
    }

    @Override
    public SessionKey getSessionKey() {
        return new SessionKeyImpl(player.getUniqueId(), player.getName());
    }

    private static class SessionKeyImpl implements SessionKey {
        // If not static, this will leak a reference

        private final UUID uuid;
        private final String name;

        private SessionKeyImpl(UUID uuid, String name) {
            this.uuid = uuid;
            this.name = name;
        }

        @Override
        public UUID getUniqueId() {
            return uuid;
        }

        @Nullable
        @Override
        public String getName() {
            return name;
        }

        @Override
        public boolean isActive() {
            // We can't directly check if the player is online because
            // the list of players is not thread safe
            return ThreadSafeCache.getInstance().getOnlineIds().contains(uuid);
        }

        @Override
        public boolean isPersistent() {
            return true;
        }
    }

    @Deprecated
    public static WorldEditPlayer wrap(EntityPlayerMP player){
        return new WorldEditPlayer(NailedPlatform.instance().getPlayerFromEntity(player));
    }

    public static WorldEditPlayer wrap(NailedPlayer player){
        return new WorldEditPlayer(player);
    }
}
